File: src/main/java/org/kite9/diagram/position/Dimension2D.java
package org.kite9.diagram.position;

/**
 * This is compatible with the awt dimension class, which is used for a
 * lot of rendering.  
 * 
 * This has double precision though and has internal scaling operations, as well
 * as actions to allow you to apply operations to a specific direction.
 * 
 * <em>Immutable</em>
 * 
 * @author robmoffat
 *
 */
public class Dimension2D {

	private double x, y;
	
	public double getHeight() {
		return y;
	}

	public Dimension2D getSize() {
		return this;
	}

	public double getWidth() {
		return x;
	}
	
	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Dimension2D) {
			Dimension2D d2d = (Dimension2D) obj;
			return (x==d2d.x) && (y==d2d.y);
		}
		return false;
	}

	public Dimension2D() {
	}
	
	public Dimension2D(double x, double y) {
		this.x = x;
		this.y = y;
	}
	
	public Dimension2D(Dimension2D clone) {
		this(clone.x, clone.y);
	}

	public Dimension2D divide(Dimension2D by) {
		Dimension2D d2 = new Dimension2D();
		d2.x = x / by.x;
		d2.y = y / by.y;
		return d2;
	}
	
	public Dimension2D multiply(Dimension2D by) {
		Dimension2D d2 = new Dimension2D();
		d2.x = x * by.x;
		d2.y = y * by.y;
		return d2;
	}
	
	public Dimension2D multiply(double by) {
		Dimension2D d2 = new Dimension2D();
		d2.x = x * by;
		d2.y = y * by;
		return d2;
	}
	
	
	public Dimension2D roundUpTo(Dimension2D factor) {
		Dimension2D d2 = new Dimension2D();
		d2.x = Math.ceil(x / factor.x) * factor.x;
		d2.y = Math.ceil(y / factor.y) * factor.y;
		return d2;
	}
	
	@Override
	public String toString() {
		return "["+x+","+y+"]";
	}

	public Dimension2D add(Dimension2D by) {
		Dimension2D d2 = new Dimension2D();
		d2.x = x + by.x;
		d2.y = y + by.y;
		return d2;
	}
	
	public Dimension2D minus(Dimension2D by) {
		Dimension2D d2 = new Dimension2D();
		d2.x = x - by.x;
		d2.y = y - by.y;
		return d2;
	}
	
	public double x() {
		return x;
	}
	
	public double y() { 
		return y;
	}

	public static Dimension2D setX(Dimension2D in, double x) {
		return new Dimension2D(x, in == null ? 0 : in.y);
	}
	
	public static Dimension2D setY(Dimension2D in, double y) {
		return new Dimension2D(in == null ? 0 : in.x, y);
	}
}

File: src/main/kotlin/org/kite9/diagram/model/position/Dimension2D.kt
package org.kite9.diagram.model.position

import kotlin.math.ceil

interface Dimension2D {

    val w: Double
    val h: Double
    fun size(): Dimension2D

    fun x(): Double {
        return w
    }

    fun y(): Double {
        return h
    }

    fun width(): Double {
        return w
    }

    fun height(): Double {
        return h
    }

    fun divide(by: Dimension2D): Dimension2D {
        val d2 = BasicDimension2D(width() / by.width(), height() / by.height())
        return d2
    }

    fun multiply(by: Dimension2D): Dimension2D {
        val d2 = BasicDimension2D(width() * by.width(), height() * by.height())
        return d2
    }

    fun multiply(by: Double): Dimension2D {
        val d2 = BasicDimension2D(width() * by, height() * by)
        return d2
    }

    fun roundUpTo(factor: Dimension2D): Dimension2D {
        val d2 = BasicDimension2D(
            ceil(width() / factor.width()) * factor.width(),
            ceil(height() / factor.height()) * factor.height()
        )
        return d2
    }

    fun add(by: Dimension2D): Dimension2D {
        val d2 = BasicDimension2D(width() + by.width(), height() + by.height())
        return d2
    }

    fun minus(by: Dimension2D): Dimension2D {
        val d2 = BasicDimension2D(width() - by.width(), height() - by.height())
        return d2
    }


    fun setY(y: Double): Dimension2D {
        return BasicDimension2D(width(), y)
    }

    fun setX(x: Double): Dimension2D {
        return BasicDimension2D(x, height())
    }
}